version: "3.8"

services:
  fluentbit:
    image: fluent/fluent-bit:2.1.10
    container_name: fluentbit
    restart: on-failure
    ports:
      - "24224:24224"
      - "24224:24224/udp"
    volumes:
      - ./scripts/fluentbit/fluent-bit.conf:/fluent-bit/etc/fluent-bit.conf
      - ./app/logs:/app/logs
    networks:
      - observability

  loki:
    image: grafana/loki:2.9.2
    container_name: loki
    restart: on-failure
    ports:
      - "3100:3100"
    volumes:
      - ./scripts/loki:/etc/loki
    command:
      - "-config.file=/etc/loki/config.yaml"
    networks:
      - observability

  grafana:
    image: grafana/grafana:latest
    container_name: grafana
    restart: on-failure
    environment:
      - GF_SECURITY_ADMIN_USER=admin
      - GF_SECURITY_ADMIN_PASSWORD=admin
      - GF_USERS_ALLOW_SIGN_UP=false
    ports:
      - "3000:3000"
    volumes:
      - ./scripts/grafana/provisioning:/etc/grafana/provisioning
      - grafana_data:/var/lib/grafana
    networks:
      - observability
  
  nginx:
    image: nginx
    container_name: nginx
    ports:
      - "8000:80"
    logging:
      driver: fluentd
      options:
        fluentd-address: fluentbit:24224
        tag: nginx
    depends_on:
      - fluentbit
    networks:
      - observability

volumes:
  grafana_data:

networks:
  observability:
    driver: bridge
